### ISSUE TITLE
Inflight error caused by larer_heap warning

### ISSUE THREAD

**Issue by moluzhui** [emqx/emqx#3645]
**Please describe your problem in detail, if necessary, you can upload the log file through the attachment**:
When I intend to forward the message to the cloud EMQX-Broker via the EMQX-Edge `bridge`, but the `large_heap warning` appears during the experiment, then the bridge is broken, EMQX-Edge is restarted, then an `Inflight error` occurs, and finally a large number of bridge messages are `sent repeatedly`, as shown in the following link

https://github.com/emqx/emqx/issues/3629#issuecomment-665401106

**Comment by moluzhui** [comment]
> **Please describe your problem in detail, if necessary, you can upload the log file through the attachment**:
> When I intend to forward the message to the cloud EMQX-Broker via the EMQX-Edge `bridge`, but the `large_heap warning` appears during the experiment, then the bridge is broken, EMQX-Edge is restarted, then an `Inflight error` occurs, and finally a large number of bridge messages are `sent repeatedly`, as shown in the following link
> 
> [#3629 (comment)](https://github.com/emqx/emqx/issues/3629#issuecomment-665401106)

I replaced it with **EMQX :v4.0.7**, and the container on the cloud can subscribe to the bridging information that EMQX-Edge occurs in a timely manner.

But EMQx-edge logs shows the following warnings and errors
```
EMQ X Edge 4.0.7 is running now!
Eshell V10.5.6  (abort with ^G)
(aa122cb83ed9@172.27.0.2)1> ['2020-08-04T13:11:48Z']:emqx start
2020-08-04 13:13:41.365 [warning] [Alarm Handler] New Alarm: system_memory_high_watermark, Alarm Info: []
(aa122cb83ed9@172.27.0.2)1>
===== ALIVE Tue Aug  4 13:28:41 UTC 2020
2020-08-04 13:31:21.714 [warning] [SYSMON] large_heap warning: pid = <0.1295.0>, info: [{old_heap_block_size,7427328},
                                             {heap_block_size,2487399},
                                             {mbuf_size,0},
                                             {stack_size,370},
                                             {old_heap_size,4461656},
                                             {heap_size,706907}]
[{initial_call,{proc_lib,init_p,5}},
 {current_function,{erlang,integer_to_list,1}},
 {registered_name,emqx_bridge_worker_aws},
 {status,running},
 {message_queue_len,2},
 {group_leader,<0.1292.0>},
 {priority,normal},
 {trap_exit,true},
 {reductions,10246317},
 {last_calls,false},
 {catchlevel,5},
 {trace,0},
 {suspending,[]},
 {sequential_trace_token,[]},
 {error_handler,error_handler},
 {memory,79318976},
 {total_heap_size,9914734},
 {heap_size,2487399},
 {stack_size,365},
 {min_heap_size,233}]
(aa122cb83ed9@172.27.0.2)1> 2020-08-04 13:31:34.251 [warning] [SYSMON] long_schedule warning: pid = <0.992.0>, info: [{timeout,325},
                                               {in,{gen_server,loop,7}},
                                               {out,{gen,do_call,4}}]
[{initial_call,{proc_lib,init_p,5}},
 {current_function,{gen_server,loop,7}},
 {registered_name,logger_disk_log_h_file},
 {status,running},
 {message_queue_len,1},
 {group_leader,<0.951.0>},
 {priority,normal},
 {trap_exit,true},
 {reductions,13408},
 {last_calls,false},
 {catchlevel,1},
 {trace,0},
 {suspending,[]},
 {sequential_trace_token,[]},
 {error_handler,error_handler},
 {memory,19308},
 {total_heap_size,2282},
 {heap_size,1598},
 {stack_size,12},
 {min_heap_size,233}]
(aa122cb83ed9@172.27.0.2)1> 2020-08-04 13:31:34.494 [warning] [SYSMON] long_schedule warning: pid = <0.1556.0>, info: [{timeout,1316},
                                                {in,
                                                 {emqx_connection,recvloop,2}},
                                                {out,
                                                 {emqx_connection,recvloop,
                                                  2}}]
[{initial_call,{proc_lib,init_p,5}},
 {current_function,{emqx_connection,recvloop,2}},
 {registered_name,[]},
 {status,waiting},
 {message_queue_len,0},
 {group_leader,<0.1051.0>},
 {priority,normal},
 {trap_exit,false},
 {reductions,12099817},
 {last_calls,false},
 {catchlevel,1},
 {trace,0},
 {suspending,[]},
 {sequential_trace_token,[]},
 {error_handler,error_handler},
 {memory,14176},
 {total_heap_size,1611},
 {heap_size,610},
 {stack_size,9},
 {min_heap_size,233}]
(aa122cb83ed9@172.27.0.2)1> 2020-08-04 13:31:43.391 [warning] [Alarm Handler] New Alarm: process_memory_high_watermark, Alarm Info: <0.1295.0>
(aa122cb83ed9@172.27.0.2)1> eheap_alloc: Cannot allocate 4723591984 bytes of memory (of type "old_heap").
heart: Tue Aug  4 13:31:55 2020: Erlang is crashing .. (waiting for crash dump file)
[os_mon] memory supervisor port (memsup): Erlang has closed
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed

=====
===== LOGGING STARTED Tue Aug  4 13:31:57 UTC 2020
=====
['2020-08-04T13:31:59Z']:emqx not running, waiting for recovery in 20 seconds
Exec: /opt/emqx/erts-10.5.6/bin/erlexec -boot /opt/emqx/releases/v4.0.7/emqx -mode embedded -boot_var ERTS_LIB_DIR /opt/emqx/erts-10.5.6/../lib -mnesia dir "/opt/emqx/data/mnesia/aa122cb83ed9@172.27.0.2" -config /opt/emqx/data/configs/app.2020.08.04.13.31.58.config -args_file /opt/emqx/data/configs/vm.2020.08.04.13.31.58.args -vm_args /opt/emqx/data/configs/vm.2020.08.04.13.31.58.args -start_epmd false -epmd_module ekka_epmd -proto_dist ekka -- console
Root: /opt/emqx
/opt/emqx
heart_beat_kill_pid = 3250
Starting emqx on node aa122cb83ed9@172.27.0.2
Start http:management listener on 8081 successfully.
Start http:dashboard listener on 18083 successfully.
Start mqtt:tcp listener on 127.0.0.1:11883 successfully.
Start mqtt:tcp listener on 0.0.0.0:1883 successfully.
Start mqtt:ws listener on 0.0.0.0:8083 successfully.
Start mqtt:ssl listener on 0.0.0.0:8883 successfully.
Start mqtt:wss listener on 0.0.0.0:8084 successfully.
EMQ X Edge 4.0.7 is running now!
Eshell V10.5.6  (abort with ^G)
(aa122cb83ed9@172.27.0.2)1> 2020-08-04 13:35:47.901 [warning] emqtt(bridge_aws): Unexpected PUBREL: 279
(aa122cb83ed9@172.27.0.2)1> 2020-08-04 13:35:47.902 [warning] emqtt(bridge_aws): Unexpected PUBREL: 280
(aa122cb83ed9@172.27.0.2)1> 2020-08-04 13:35:47.902 [warning] emqtt(bridge_aws): Unexpected PUBREL: 281
(aa122cb83ed9@172.27.0.2)1> 2020-08-04 13:35:47.902 [warning] emqtt(bridge_aws): Unexpected PUBREL: 282
(aa122cb83ed9@172.27.0.2)1> 2020-08-04 13:35:47.903 [warning] emqtt(bridge_aws): Unexpected PUBREL: 283
(aa122cb83ed9@172.27.0.2)1> 2020-08-04 13:35:47.903 [warning] emqtt(bridge_aws): Unexpected PUBREL: 284
(aa122cb83ed9@172.27.0.2)1> 2020-08-04 13:35:47.904 [warning] emqtt(bridge_aws): Unexpected PUBREL: 285
```

There is still a `large_heap` alarm, but the message occurs in time. Not sure why this is the case

**Comment by HJianBo** [comment]
Based on your description, I think the current issue falls into several parts:

1. The `emqx-edge` container crash frequently:
> eheap_alloc: Cannot allocate 4723591984 bytes of memory (of type "old_heap").
heart: Tue Aug  4 13:31:55 2020: Erlang is crashing .. (waiting for crash dump file)
[os_mon] memory supervisor port (memsup): Erlang has closed
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed

**It is an OOM (Out of memory) crash. It probably has a large number of messages piled up in the emqx-edge** (That's why the `large_heap_size` warning was generated.).

Could you share me your test scenario in the this testing? Is it similarly with https://github.com/emqx/emqx/issues/3629#issuecomment-665401106 ?:
```
EMQX_BRIDGE__MQTT__AWS__ADDRESS=xx.xxx.xxx.xxx:1883
EMQX_BRIDGE__MQTT__AWS__START_TYPE=auto
EMQX_BRIDGE__MQTT__AWS__FORWARDS=data/321/#
EMQX_BRIDGE__MQTT__AWS__FORWARD_MOUNTPOINT=/
EMQX_BRIDGE__MQTT__AWS__SUBSCRIPTION__1__TOPIC=cmd/123/#
EMQX_BRIDGE__MQTT__AWS__SUBSCRIPTION__1__QOS=2
EMQX_BRIDGE__MQTT__AWS__CLEAN_START=false

docker run -d -p 1883:1883 --env-file env.list --name edgeemqx411 emqx/emqx-edge:v4.1.1
```

If this is the case, **can you estimate the message throughput from EMQX-EDGE to EMQX-BROKER and from EMQX-BROKER to EMQX-EDGE? (E.g: the number of messages per second, and the payload size, Qos level of messages)**


2. Due to the EMQX-EDGE  subscribing a topics  to the emqx-broker as Qos2 and `CLEAN_START=false`:
```
EMQX_BRIDGE__MQTT__AWS__SUBSCRIPTION__1__TOPIC=cmd/123/#
EMQX_BRIDGE__MQTT__AWS__SUBSCRIPTION__1__QOS=2
EMQX_BRIDGE__MQTT__AWS__CLEAN_START=false
```

After the EMQX-EDGE restarted**, the EMQX-BROKER will resend PUBLISH and PUBREL packet to `EMQX-EDGE`. There should be a BUG here**, so the EMQX-EDGE and EMQX-BROKER reported `emqtt(bridge_aws): Unexpected PUBREL: 279` and `[Channel] The PUBREC PacketId 1233 is inuse.` errores.

Can you use QoS1 to subscribe to Broker's topics? It gets around these kinds of problems, but it can lead to duplication of messages? 

**We will try to fix it at next version**

3. These logs won't have any additional impact. We have fixed it at v4.0.4 (see: https://github.com/emqx/emqx/pull/3290/files), **But now it's still happening in 4.0.7, We'll check again later.**
```
Unexpected sock_closed: einval
```

**Comment by HJianBo** [comment]
@tigercl Please check the Problem 3
